website/project/views/comment.py

```diff
@@ -17,30 +18,69 @@
+            provider=source['provider'],
+            guids=[],
+            node=source_node)
+
+        if event_type == 'addon_file_renamed' and (source['provider'] == 'osfstorage' or source['provider'] == 'box'):
+            return
+        if event_type == 'addon_file_moved' and (source['provider'] == destination['provider'] and (source['provider'] == 'osfstorage' or source['provider'] == 'box')) and source_node == destination_node:
+            return
+
+        for guid in file_guids:
+            obj = Guid.load(guid)
+            if source_node != destination_node and Comment.find(Q('root_target', 'eq', guid)).count() != 0:
+                update_comment_node(guid, source_node, destination_node)
+
+            if not (source['provider'] == destination['provider'] and source['provider'] == 'osfstorage'):
+                if not source['path'].endswith('/'):
+                    data = dict(destination)
+                    new_file = FileNode.resolve_class(destination['provider'], FileNode.FILE).get_or_create(destination_node, destination['path'])
+                    if destination['provider'] != 'osfstorage':
+                        new_file.update(revision=None, data=data)
+                else:
+                    new_file = create_file_from_metadata(destination.get('children', []), source, destination, destination_node, obj)
+                    if not new_file:
+                        if source['provider'] == 'box':
+                            new_path = obj.referent.path
+                        else:
+                            new_path = obj.referent.materialized_path.replace(source['materialized'], destination['materialized'])
+                        new_file = FileNode.resolve_class(destination['provider'], FileNode.FILE).get_or_create(destination_node, new_path)
+                        new_file.name = obj.referent.name
+                        new_file.materialized_path = new_path
+                        new_file.save()
+
+                obj.referent = new_file
+                obj.save()
+
+
+def create_file_from_metadata(children, source, destination, destination_node, obj):
     for item in children:
-        if not item.is_file:
-            update_folder_contents(item.children, source_node, destination_node)
-        else:
-            Comment.update(Q('root_target', 'eq', item._id), data={'node': destination_node})
-            # update node record of commented files
-            if item._id in source_node.commented_files:
-                destination_node.commented_files[item._id] = source_node.commented_files[item._id]
-                del source_node.commented_files[item._id]
-                source_node.save()
-                destination_node.save()
+        if item['kind'] == 'folder':
+            return create_file_from_metadata(item.get('children', []), source, destination, destination_node, obj)
```

Comment:
This is recursive? Could you add a docstring describing what it's doing and why? I'm guessing it's for… moving a folder or file and ensuring you keep the comments?